camus-api/src/main/java/com/linkedin/camus/schemaregistry/CachedSchemaRegistry.java

```diff
@@ -1,87 +1,156 @@
+  private static final String GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS_DEFAULT = "1";
+
+  private final SchemaRegistry<S> registry;
+  private final ConcurrentHashMap<CachedSchemaTuple, S> cachedById;
+  private final ConcurrentHashMap<String, SchemaDetails<S>> cachedLatest;
+  private int getByIdMaxRetries;
+  private int getByIdMinIntervalSeconds;
+  private Map<String, FailedFetchHistory> failedFetchHistories;
+
+  private class FailedFetchHistory {
+    private int numOfAttempts;
+    private long previousAttemptTime;
+
+    private FailedFetchHistory(int numOfnumOfAttempts, long previousAttemptTime) {
+      this.numOfAttempts = numOfnumOfAttempts;
+      this.previousAttemptTime = previousAttemptTime;
+    }
+  }
+
+  public void init(Properties props) {
+    this.getByIdMaxRetries =
+        Integer.parseInt(props.getProperty(GET_SCHEMA_BY_ID_MAX_RETIRES, GET_SCHEMA_BY_ID_MAX_RETIRES_DEFAULT));
+    this.getByIdMinIntervalSeconds =
+        Integer.parseInt(props.getProperty(GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS,
+            GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS_DEFAULT));
+  }
+
+  public CachedSchemaRegistry(SchemaRegistry<S> registry, Properties props) {
+    this.registry = registry;
+    this.cachedById = new ConcurrentHashMap<CachedSchemaTuple, S>();
+    this.cachedLatest = new ConcurrentHashMap<String, SchemaDetails<S>>();
+    this.failedFetchHistories = new HashMap<String, FailedFetchHistory>();
+    this.init(props);
+  }
+
+  public String register(String topic, S schema) {
+    return registry.register(topic, schema);
+  }
+
+  public S getSchemaByID(String topic, String id) {
+    CachedSchemaTuple cacheKey = new CachedSchemaTuple(topic, id);
+    S schema = cachedById.get(cacheKey);
+    if (schema == null) {
+      synchronized (this) {
+        if (shouldFetchFromSchemaRegistry(id)) {
+          schema = fetchFromSchemaRegistry(topic, id);
+        } else {
+          throw new SchemaNotFoundException();
+        }
+      }
+    }
+    return schema;
+  }
+
+  private boolean shouldFetchFromSchemaRegistry(String id) {
+    if (!failedFetchHistories.containsKey(id)) {
+      return true;
+    }
+    FailedFetchHistory failedFetchHistory = failedFetchHistories.get(id);
+    boolean maxRetriesNotExceeded = failedFetchHistory.numOfAttempts < getByIdMaxRetries;
+    boolean minRetryIntervalSatisfied =
+        System.nanoTime() - failedFetchHistory.previousAttemptTime >= TimeUnit.SECONDS
+            .toNanos(getByIdMinIntervalSeconds);
+    return maxRetriesNotExceeded && minRetryIntervalSatisfied;
+  }
+
+  private S fetchFromSchemaRegistry(String topic, String id) {
+    try {
+      S schema = registry.getSchemaByID(topic, id);
+      cachedById.putIfAbsent(new CachedSchemaTuple(topic, id), schema);
+      return schema;
+    } catch (SchemaNotFoundException e) {
+      addFetchToFailureHistory(id);
+      throw e;
+    }
+  }
+
+  private void addFetchToFailureHistory(String id) {
```

Comment:
Add a javadoc note indicate that caller needs to do the synchronization.

On the other hand, it probably doesn't hurt to have 'synchronized' here since the caller already holds the same lock.